### ISSUE TITLE
[WIP] [RFC] Distributed Correctness Testing Framework

### ISSUE THREAD

**Issue by Swiddis** [opensearch-project/sql#3220]
# [WIP] [RFC] Distributed Correctness Testing Framework
**Date:** Dec 23, 2024
**Status:** Work In Progress

## Overview

While the current integration testing framework is effective for development, there have been several bug reports around query correctness due to testing blind-spots. Some common sources for issues include:

- Using complex data types such as [arrays](https://github.com/opensearch-project/sql/issues/3132), [multi-typed fields](https://github.com/opensearch-project/sql/issues/1074), [nested fields](https://github.com/opensearch-project/opensearch-spark/issues/233), or [timestamp fields](https://github.com/opensearch-project/sql/issues/1686).
- Queries involving multiple indices such as [`JOIN`s](https://github.com/opensearch-project/sql/issues/2550) or [wildcard indices](https://github.com/opensearch-project/sql/issues/1542).
- Interactions between our plugin and supported external interfaces such as [JDBC](https://github.com/opensearch-project/sql/issues/1874) and [Spark](https://github.com/opensearch-project/opensearch-spark/issues/739).
- Running queries with large datasets such as those involving [large documents](https://github.com/opensearch-project/opensearch-spark/issues/625), [high-cardinality data](https://github.com/opensearch-project/opensearch-spark/issues/740), as well as associated [`LIMIT` queries](https://github.com/opensearch-project/sql/issues/3121) or [aggregations](https://github.com/opensearch-project/sql/issues/2716).

In addition to the obvious challenges with predicting these edge cases ahead of time, we have the additional issue that the SQL plugin has multiple runtime contexts:

- Different query engines such as Legacy, [V2](https://github.com/opensearch-project/sql/blob/main/docs/dev/intro-v2-engine.md), [Spark](https://github.com/opensearch-project/opensearch-spark), or (soon) Catalyst.
- [Several configuration options](https://github.com/opensearch-project/sql/blob/3430deb57514a4c593813e8735d62484756c840d/common/src/main/java/org/opensearch/sql/common/setting/Settings.java#L21-L62) that have effects on the runtime behavior of queries (e.g. pagination mode, size or memory limits, type tolerance).
- Supporting both SQL and PPL querying.
- Index-specific settings such as [shards](https://github.com/opensearch-project/sql/issues/3064) or [date formats](https://github.com/opensearch-project/sql/issues/2700).

The current integration testing process is based on doesn't scale sufficiently to detect edge cases under all of these scenarios. Historically the method has been "fix the bugs as we go", but given the scale of the project nowadays and in-progress refactoring initiatives, a testing suite that can keep up with the scale is needed.

Inspired by the likes of database testing initiatives in other projects, especially [SQLancer](https://github.com/sqlancer/sqlancer), [Google's OSS-Fuzz](https://github.com/google/oss-fuzz), and the [PostgreSQL BuildFarm](https://buildfarm.postgresql.org/), this RFC proposes a project to implement a distributed random testing process, which will be able to generate a large amount of test scenarios to validate correct behavior. With such a suite, the plugin can be "soak tested" by running a large number of randomized tests and reporting errors.

### Glossary

- **Crash Bugs:** Bugs that are directly visible as erroneous behavior (e.g. error responses, crashes). Distinct from Logic Bugs.
- **Fuzzing:** A testing technique that involves providing invalid, unexpected, or random data as inputs to a system. This usually has a heavier emphasis on asserting the system doesn't crash, as opposed to making sure the responses are actually correct.
- **Logic Bugs:** Bugs where the system does not have an obviously incorrect behavior (e.g. error responses, crashes), but nonetheless returns incorrect results. Distinct from Crash Bugs.
- **Property-based testing:** A testing methodology that focuses on verifying properties of the system under test for a wide range of randomly-selected inputs. Further reading: [What is Hypothesis?](https://hypothesis.works/articles/what-is-hypothesis/) and [ScalaCheck](https://scalacheck.org/documentation.html).

## Current State

Our current testing infrastructure consists of several components that serve different purposes, but none of them clearly address the gaps identified here.

1.  The Main Integration Testing Suite:
    
- Provides comprehensive coverage for current features, individually.
- Effective for catching regressions in existing functionality.
- Limited in its ability to detect edge cases or cross-feature interactions.

2.  Comparison Testing Suite:
    
- Attempts to validate query results by comparing against SQLite.
- Useful for basic SQL functionality shared between our implementation and SQLite.
- Limited by significant differences in feature sets:
    - Cannot test OpenSearch-specific features (e.g. complex data types, PPL).
    - May produce false positives due to intentional differences in behavior.
- These tests have not actually been run in a very long time, and seem to currently be out of commission.

We may consider reusing elements of the comparison testing framework for the new suite. In particular, both this framework and the proposed solution connect to OpenSearch via JDBC. The main concern is whether we can parallelize the workload easily.

3.  Mutation Testing Suite (PiTest):
    
- Helps validate the effectiveness of our existing test suite under code mutations.
- Identifies areas where our tests may be insufficient.
- Still constrained by the scope of our existing tests, as it doesn't generate new test cases.
- These tests also seem to currently be out of commission.

While these tools provide valuable testing capabilities, they fall short in several key areas:

- Limited ability to generate diverse, representative test data.
- Insufficient coverage of complex interactions between different SQL and PPL features.
- Difficulty in testing across multiple runtime contexts and configurations.
- Challenges in scaling to test large datasets or high-volume query scenarios.

## Analysis of Domain

The most important question to ask is, "why can't we use an existing system?" SQL testing in particular has a lot of existing similar initiatives. A cursory look at these initiatives (ref: [SQLancer](https://github.com/sqlancer/sqlancer), [Sqllogictest](https://www.sqlite.org/sqllogictest/doc/trunk/about.wiki), [SQuaLity](https://arxiv.org/abs/2410.21731)) reveals a few issues:

- These suites aren’t designed with PPL or generally non-SQL query languages in mind, so the work to migrate their code to support both languages would be significant.
- These suites generally assume that the database is not read-only: they include detailed tests for modifying database files directly (including corrupting them), as well as several tests that depend on the behavior of SQL’s write functions (`INSERT`, `UPDATE`). Removing this functionality and replacing it with a system that can set up a correctly-configured OpenSearch cluster would be again similar to writing the system from scratch.
- Since our SQL implementation has several limitations that aren’t shared by other systems, we also would need to remove tests that rely on these limitations not being present. This is nontrivial since many of these systems rely on generating queries in bulk, and may implicitly use those same assumptions even for functionality we support.

Compared to trying to adapt these solutions, it seems like the most reliable method for a long-term solution is to write a custom system. In particular, these methods are well-documented, so we likely can make something that can get a similar degree of effectiveness with less effort than the migration cost. This will also give us a lot of flexibility to toggle specific features under test, such as the assortment of unimplemented features in V2 (#1889, #1718, #1487). The flexibility of supporting OpenSearch-specific semantics and options will also open routes for implementing similar testing on core OpenSearch.

Despite these limitations, all of the linked projects have extensive research behind them that can be used to guide our own implementation. Referencing them and taking what we can is likely valuable.

## Goals

Looking at the limitations of our current test cases, I propose the following goals:

- **Scalability:** Currently, it's nontrivial to run our integration tests in parallel without multiple clusters due to cross-test interference. To run the tests at the scale necessary for these types of suites to find deep bugs, we need to ensure test isolation from the start. This ties in with the next goal:
- **Cost Effectiveness:** The costs of the suite should directly correlate with the amount of tests we're running. For the most part, running clusters is the largest cost, so the suite should be able to load balance parallelized tests across a small number of clusters.
- **Model Real Usage:** The existing suites use a small amount of test indices for specific features. The indices generated for the tests should more closely model real indices.

## Implementation Strategy

[WIP] have a rough outline, but not a high-fidelity design.

The rough outline is: to create a separate project that can connect to OpenSearch SQL via the JDBC connector -- similar to the comparison testing. We use the JDBC connector because it will let us use ScalaCheck to write the properties, working well with our current development ecosystem. We'll use that connector to run queries on test-individual indices (for isolation). At first we will only generate simple queries, but gradually extend it to match newly introduced features and cover existing features. The tests will more-or-less run independently in the background (likely on EC2), as part of a cluster separate from GitHub actions, and report bugs asynchronously as we make changes. Common bug types can get specialized tests introduced in the main test suite.

Tests will follow these steps:
- Configure a cluster. For each cluster:
- Generate data sets, each containing (potentially multiple) indices. For each data set:
- Generate (read-only) queries for that data set, including multiple languages.
- Validate query results on the data set in parallel.
- Once the queries are done, delete their data set. Once all data sets are done, stop the cluster and repeat.

For properties we can test, the following high-level strategies are available:
- Property-based testing as a testing methodology is described in the glossary.
- Pivoted Query Synthesis (PQS), Non-Optimizing Reference Engine Construction (NoREC), Ternary Logic Partitioning (TLP), Query Plan Guidance (QPG), and Differential Query Plans (DQP) are all described in the [SQLancer Papers](https://github.com/sqlancer/sqlancer/blob/main/docs/PAPERS.md). (TODO: will add descriptions/examples of all of these in this doc.)
- SQuaLity also has some techniques aggregated from other database systems, but I'm waiting to hear back on https://github.com/suyZhong/SQuaLity/issues/26.

Some open questions:
- Does JDBC actually support PPL? If not, how do we plan on supporting PPL?
- Where will testing results be reported?
- How will we handle versioning/compatibility across OpenSearch versions?
- What metrics will we use to measure the effectiveness of this new testing framework compared to our current testing methods?
- What mechanisms will we put in place to ensure that the generated test cases are reproducible when bugs are found?
- How will we handle testing of queries that involve very large result sets or long-running operations?
- How will we handle testing of queries that involve external resources (e.g., SQL functions that call external APIs)? Spark is the primary example here.
- How will we handle testing of security features and access controls within this framework?
- How will we prioritize which features or combinations of features to test first? (Likely those under active development.)
- What provisions will we make for testing query behavior under different cluster states (e.g., node failures, network partitions)?

**Comment by RyanL1997** [comment]
Hi @Swiddis , thanks for putting this together. I know some of the parts are still work in progress, so just wanna leave some general thoughts after reading the current version as my first round of review:

1. "For the edge cases listed in the overview section, is there a severity or priority ranking to help us identify the most critical issues to address first? Or is the intention to resolve all of them simultaneously? This would provide better clarity on the focus and scope of the implementation."

2. > Interactions between our plugin and supported external interfaces such as JDBC and Spark.

I saw that you also mentioned in the open questions: "How will we handle versioning/compatibility across OpenSearch versions?" Especially for these external dependencies, I'm actually having the similar question that will backward compatibility testing be a focus as part of the proposed framework? Are there plans to define specific test cases or scenarios to validate compatibility with older versions, and how will this be managed over time as new features are introduced?

3. > Challenges in scaling to test large datasets or high-volume query scenarios.

Personally, I need to do more homework on the scenario for large datasets. However, by taking a first look, I think It might be useful to include a plan for benchmarking these scenarios to measure performance impact. (This may not be the P0)

4. > The tests will more-or-less run independently in the background (likely on EC2), as part of a cluster separate from GitHub actions, and report bugs asynchronously as we make changes. Common bug types can get specialized tests introduced in the main test suite.

I noticed the mention of the above. Does this imply the introduction of another internal test pipeline, distinct from the open-source GitHub-based pipeline? If so, how will the results from this EC2-based pipeline be integrated or communicated back to the broader testing framework?

**Comment by RyanL1997** [comment]
Also, for this open question: 

> What metrics will we use to measure the effectiveness of this new testing framework compared to our current testing methods?

Here are some general ideas in my mind:

**Coverage Metrics**
- Test Coverage Improvement: Measure the percentage of edge cases, scenarios, or code paths covered by the new framework compared to the current one.
- Edge Case Handling: Number of previously missed edge cases or bugs identified and addressed by the new framework.

**Accuracy and Reliability Metrics**
- False Positive Rate: Track reductions in the rate of incorrect bug reports generated by the framework.
- Bug Detection Accuracy: Percentage of valid bugs detected out of the total issues flagged by the framework.

**Efficiency Metrics**
- Execution Time: Compare the time taken to execute tests in the new framework versus the current one.
- Test Resource Utilization: Assess compute resource usage (e.g., EC2/GH runner?) of the new testing framework versus existing pipelines.

**Outcome-Based Metrics**
- Post-Release Bug Rate: Measure the reduction in bugs reported after release due to better pre-release testing.
- Severity of Undetected Bugs: Evaluate the severity of any bugs that still slip through the framework compared to before.

**Integration and Usability Metrics**
- Integration Time: Time required to integrate the new framework into existing pipelines or workflows.
- Ease of Debugging: Feedback from developers on how easily the framework helps identify and resolve bugs.

**Comment by Swiddis** [comment]
I tried setting up some crude benchmarking code to estimate how much test throughput we could get. I tried versions of a [Locust](https://locust.io/) script that tried two scenarios. Left: many parallel tests that each create, update, and delete their own index. Right: many tests sharing a small number of indices.

<img src="https://github.com/user-attachments/assets/37edf866-0ab8-4c4c-b5eb-5644aabef206" width="400">
<img src="https://github.com/user-attachments/assets/3ad6210a-381e-42d3-a87a-1c224d10ec5b" width="400">

In general, insert and query requests are very fast while index create/delete is slow.

- Index-per-test
![image](https://github.com/user-attachments/assets/f0e530e7-82e9-4f73-bd88-0a2421586945)

- Shared indices:
![image](https://github.com/user-attachments/assets/d6d99788-ac6b-4a0f-9667-f36f1bc1d33e)

This gives some data to back up that we should have batches of tests run on batch-scoped indices, it's a throughput difference of 25x on equal hardware. It's also more reliable -- with a small number of test-scoped index workers there were timeout failures with even just 50 tests in parallel, but batched workers can handle 1000 tests in parallel without any failures (cluster generally just slows to land at around 1800 requests/second on my machine). For a real soak-test we probably want to run at least O(a million) tests total, which my dev machine can do for this benchmark in ~10 minutes.

**Comment by Swiddis** [comment]
@RyanL1997 

> For the edge cases listed in the overview section, is there a severity or priority ranking to help us identify the most critical issues to address first? Or is the intention to resolve all of them simultaneously? This would provide better clarity on the focus and scope of the implementation.

I think we can work backwards from specific past bug reports (such as those linked in the overview) to the features the query generator needs to support, then see if we reproduce them. If the process is able to find specific known issues, we can have some confidence in its ability to find more unknown issues as we add more features.

> I saw that you also mentioned in the open questions: "How will we handle versioning/compatibility across OpenSearch versions?" Especially for these external dependencies, I'm actually having the similar question that will backward compatibility testing be a focus as part of the proposed framework?

I think we should leave it at first. Since we don't typically update older versions, running tests there would mostly be data collection. I do know there have been some bugs involving specific upgrade flows ([example](https://github.com/opensearch-project/dashboards-reporting/issues/371)), but I'm not convinced the extra complexity would be paid back.

> Personally, I need to do more homework on the scenario for large datasets. However, by taking a first look, I think It might be useful to include a plan for benchmarking these scenarios to measure performance impact.

We can probably extend the testing to do some sort of benchmarking, but let's not step on the toes of [opensearch-benchmark](https://github.com/opensearch-project/opensearch-benchmark) where we don't have to. The focus for the moment is just correctness. Since we'll be running many tests per index (see above comment), we can probably afford to make each index much larger than if we were doing 1 test per index.

> Does this imply the introduction of another internal test pipeline, distinct from the open-source GitHub-based pipeline? If so, how will the results from this EC2-based pipeline be integrated or communicated back to the broader testing framework?

It will probably be some sort of email mailing list, or a dashboard that's periodically checked. We shouldn't automatically publish issues to GH (both to avoid noise and also in case there's any issues that are security-relevant). That said, I think my original idea of running this 24/7 and just tracking failures also is complicated, it'd require a whole live notification system. For simplicity, let's make the suite run a configurable finite number of tests with an existing framework, like how Hypothesis does [`max_examples`](https://hypothesis.readthedocs.io/en/latest/settings.html#hypothesis.settings.max_examples), and deposit the HTML report somewhere. We can upgrade it if we need to, I'm not convinced we will.

**Comment by anasalkouz** [comment]
Thanks @Swiddis for putting this proposal, I have few generic comments:

> These suites generally assume that the database is not read-only: they include detailed tests for modifying database files directly (including corrupting them), as well as several tests that depend on the behavior of SQL’s write functions (INSERT, UPDATE). Removing this functionality and replacing it with a system that can set up a correctly-configured OpenSearch cluster would be again similar to writing the system from scratch.

If this is the main concern of using out the box existing frameworks, can we explore the option to provide those write functions if possible, even if those only for testing purposes.

> - Generate data sets, each containing (potentially multiple) indices (and accelerations for Spark). For each data set:
> - Generate (read-only) queries for that data set, including multiple languages.

How to make sure we keep data sets and queries up to date with our new development on SQL/PPL language? since those most probably will be on a separate repository

**Comment by Swiddis** [comment]
@anasalkouz 

> If this is the main concern of using out the box existing frameworks, can we explore the option to provide those write functions if possible, even if those only for testing purposes.

It's possible, but I'm not sure it's efficient. One way that could work is to have these frameworks create a SQL table locally, and we write some logic that transforms the SQL table into an OS index before starting the read queries. Getting all the datatypes to work right would be tricky, and we also would still face the other issues with PPL and OpenSearch-specific semantics.

To clarify: I think these frameworks do have valuable lessons to teach, and I think we would benefit a lot from copying select parts of their code where we can. I just think that building something from scratch will give us flexibility that we can't really have otherwise.

> How to make sure we keep data sets and queries up to date with our new development on SQL/PPL language? since those most probably will be on a separate repository

So far the best idea I have is making it a reviewer step, reviewers should enforce whether we need to add test features for the new functionality[^1]. This is already done for OSD with the functional test repository.

[^1]: Not strictly related to this RFC: to mechanize this, I think we would benefit a lot from introducing a full-featured "reviewer checklist" that specifies specific things to check for. We already have PR checklists for PR authors but they're mostly ignored. Having a reviewer checklist is a practice that seems pretty common for other organizations and may reduce review churn a lot for both authors and reviewers.

**Comment by Swiddis** [comment]
I updated the document with an implementation strategy and diagrams, and left answers for several of the open questions I had (and removed a few that felt like they weren't relevant anymore). Going to call this ready for review, a lot of the remaining grungy specifics of the implementation can be done when we actually start writing.

**Comment by YANG-DB** [comment]
@Swiddis  Its a Great review and document !

### My comments regarding the framework's `Goals`:
I would like to add the next goals or expand the `Model Real Usage` goal:

 - Allow vendor specific compatibility tests - for example with Spark based PPL
     - `Standard Spark based PPL Tests`
     - `EMR Spark based PPL Tests`
     - Add non trivial datasource tests based validations (`S3 / parquet / json` ) for SQL/PPL correctness validation
     
 These  tests should be simple and easy to run using `docker-compose` or other containerized based solution 

 - Allow Performance benchmark & testing as part of this framework capabilities 
    - create multiple levels of V^3 dimensions based testing branches:
       - Volume  - amount of data
       - Velocity - speed of data being ingested ( window streaming )
       - Variety   - test multiple data types both structured and un-structured
       -

**Comment by penghuo** [comment]
@Swiddis Great document. Couple comments

> To clarify: I think these frameworks do have valuable lessons to teach, and I think we would benefit a lot from copying select parts of their code where we can. I just think that building something from scratch will give us flexibility that we can't really have otherwise.

Q1, I assume we aim to build new frameworks to address scalability issues. Since OpenSearch core integration tests currently take 2-3 hours to run, have we considered improving the existing OpenSearch integration test framework as well?
Could you elaborate more on "batches of tests run on batch-scoped indices"? Does the current integration test framework already support this capability?

Q2, Could eleberate more on Query Generator and the Data Generator?
Q3, For Spark-SQL/PPL test, we do not need to depend on OpenSearch, right?

**Comment by Swiddis** [comment]
> Allow vendor specific compatibility tests - for example with Spark based PPL: Standard Spark based PPL Tests, EMR Spark based PPL Tests

> Could you elaborate more on the Query Generator and the Data Generator?

As a goal I think that configuration of generated query features needs to be taken into account relatively early, so there will be some config that lets one enable/disable certain features. I'll be adding more details on query generation soon that should cover this. But what are the practical differences between standard Spark and EMR Spark that the test suite needs to check?

As a sketch:

We can represent query syntax trees as a sort-of [definite clause grammar (DCG)](https://www.metalevel.at/prolog/dcg), and separately implement ways to serialize those trees, maybe based on [serde serialization](https://www.joshmcguigan.com/blog/understanding-serde/) which is known to pretty heavily generalize to different serialization formats. To make sure generated queries make semantic sense, the generator is also provided with information about the test indices (columns and data types).

"Toggling features" can be as simple as enabling/disabling certain branches of the DCG. It's something I've done before while working with parsing in Prolog, and I've experimented with it for query generation before to decent success as well, but I'm not sure how well it scales for SQL and PPL specifically. (At least SQLancer seems to follow a similar approach, splitting up generator rules [by semantic function](https://github.com/sqlancer/sqlancer/tree/main/src/sqlancer/sqlite3/ast) and generating trees from those functions with a visitor.)

> Since OpenSearch core integration tests currently take 2-3 hours to run, have we considered improving the existing OpenSearch integration test framework as well? Could you elaborate more on "batches of tests run on batch-scoped indices"? Does the current integration test framework already support this capability?

This is what I was trying to look at with the benchmarking earlier -- in our current integration tests, we have many suites that each have a setup step of creating indices, run some number of tests on the indices, and tear down the indices. This limits the speed of our tests considerably:

- Creating and deleting indices is ~100x slower in bulk than regular search/update operations, so this bottlenecks the top speed. This ignores indexing time for adding sample data, which will add even more overhead.
- We can't parallelize test suites because many suites rely on the same indices -- two suites can't use the "accounts" dataset at once, for instance.

To work within these constraints, a "batch-scoped index" means that for each batch of N queries, we are going to create a new index with a randomized (probably UUID) name.

- We don't see the create/delete bottleneck as much as we have a much higher query/index ratio
- We can trivially parallelize more batches as now different batches are guaranteed to not share any indices

We could also consider doing something like this for the current integration testing. We need to split tests into read-only tests and tests that involve index writes, and then we can just set up all the test indices once to run every read-only suite before tearing it down. I'm not sure how much work it would take to convert all of them like that.

**Comment by Swiddis** [comment]
I'm working on getting a proof-of-concept starting point spun up. V0.1 of the POC will do 6 things:

- [ ] Run property batches that each define a batch-scoped index
- [ ] Connect to a hardcoded localhost cluster and create the indices
- [ ] Populate indices with sample data
- [ ] Generate simple read-only queries in bulk and run them within each batch
- [ ] Check a simple property per-query
- [ ] Generate a result report

The steps here shouldn't clash with any alternatives as long as we agree on the core idea of having the test suite be a standalone project that can be run next to some cluster. I think having the POC will give a lot more ground to flesh out answers to current open questions. ETA end of week to have decent-enough ground on all of these.

**Comment by Swiddis** [comment]
WIP @ https://github.com/Swiddis/opensearch-sql-distributed-testing

Current status:
- Have some basic cluster connection logic, will be relying on that until we decide supporting multiple cluster configs in a single runner is necessary (even then, maybe we should just have multiple runners?)
- Some very rudimentary query generation, still need to work out a lot of grammar stuff, but the data model for QueryContext is underway
- Using a sample index right now for testing, will add index creation in the future but for now having ~4.5k rows of sample data isn't too bad
- Haven't encoded any of the fancier properties, currently just "limit N returns N rows", which thankfully passes
- Tried experimenting with generating queries [from the ANTLR grammar directly](https://github.com/renatahodovan/grammarinator), but the results are too random to be useful.

That's enough of a skeleton to feel ready to push publicly, but I want to get a bit more functionality in before trying to ask for an org repository.

At the time I'm not sure what the biggest unknown is. Expanding the query features at this stage is a lot of grungy string manipulation. Most of my slowness right now is more due to my unfamiliarity with Scala than with difficulty with the concepts (I've never done an end-to-end project in it before).

In the short term, I think I'm going to try and pick an existing bug and see if I can write a generator that reproduces it, and then expand from there.

**Comment by Swiddis** [comment]
Current status: have some basic properties running now, and useful results are being generated (#3266, #3272). Adding more sophisticated tests are blocked by these failures, since they both are pretty significant wedges in having the ability to "just run queries." In general I think the future of this project is continuing to add functionality gradually and fixing bugs as they come up, and maybe (longer term) working to add in other plugins that want to contribute.

The next steps are:
- [ ] Add some documentation for the project -- how it works, how to run it, important components, etc.
- [ ] Implement enough of the grammar to cover some known bug reports, giving confidence that the approach can scale and adding more query features is sufficiently easy.
- [ ] Implement actual correctness properties -- currently this is blocked by 400s raised from the current parsing system, so we can only (usefully) test against crashes at the moment. This probably means getting to a point where the current tests are passing by patching the bugs found above.
- [ ] Get the code running in some sort of pipeline that can keep producing results -- this can also wait until after the current/extended system is passing individual runs. No real point in setting up complex infrastructure when failures are being reliably found in <10 tests.
- [ ] Find a home for the code. Likely this means a new repo instead of merging it into this one, because that affords more flexibility with testing other plugins in the future, and has less complexity with getting it running independently. I've been having some conversations with people on the indexing team about extending the work to cover some of their work too.

**Comment by Swiddis** [comment]
I'm renaming the project from "distributed testing" to "dynamic testing," because in hindsight I think putting the emphasis on running the tests at scale was the wrong approach. It's really more about writing a testing suite that automatically finds buggy cases, instead of just trying to be a "very big test suite."

**Comment by Swiddis** [comment]
Just an update on this since it's been a while: project's currently stalled while waiting for fixes to #3317, #3266, #3272, #3273 which are difficult to work around from a query generation perspective. Some of them will be fixed by Calcite, others need ANTLR updates.

**Comment by Swiddis** [comment]
Closing in favor of initiatives like #5505 and #5598
